server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java

```diff
@@ -391,29 +431,131 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         return builder;
     }
 
+    public static Set<Role> BUILT_IN_ROLES = Set.of(MasterRole.INSTANCE, DataRole.INSTANCE, IngestRole.INSTANCE);
+    private static Map<String, Role> roleNameToPossibleRoles;
+
+    public static void setPossibleRoles(final Set<Role> possibleRoles) {
+        final Map<String, Role> roleNameToPossibleRoles =
+                possibleRoles.stream().collect(Collectors.toUnmodifiableMap(Role::roleName, Function.identity()));
+        // collect the abbreviation names into a map to ensure that there are not any duplicate abbreviations
+        final Map<String, Role> roleNameAbbreviationToPossibleRoles = roleNameToPossibleRoles.values()
+                .stream()
+                .collect(Collectors.toUnmodifiableMap(Role::roleNameAbbreviation, Function.identity()));
+        assert roleNameToPossibleRoles.size() == roleNameAbbreviationToPossibleRoles.size() :
+                "roles by name [" + roleNameToPossibleRoles + "], roles by name abbreviation [" + roleNameAbbreviationToPossibleRoles + "]";
+        DiscoveryNode.roleNameToPossibleRoles = roleNameToPossibleRoles;
+    }
+
+    public abstract static class Role {
```

Comment:
I think there is sufficient substance here with the builtin roles that this could be moved out to it's own source file, rather than adding 100 lines here. Some basic javadocs would also be nice since this is now an officially pluggable api.

Reply:
Sure, I’ll do all of that.